### PR TITLE
Add source code for Chapter 8

### PR DESCRIPTION
**Pull request by Peter-VanWeert**

### CODE REVIEW

File: Examples/NoModules/Chapter 08/Ex8_06.cpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+            << std::endl;
+}
+
+// Function to compute total yield
+double yield(const double array[][4], size_t size)
+{
+  double sum  {};
+  for (size_t i {}; i < size; ++i)   // Loop through rows
+  {
+    for (size_t j {}; j < std::size(array[i]); ++j) // Loop through elements in a row
```
Comment (MarcGregoire) [inline Examples/NoModules/Chapter 08/Ex8_06.cpp:23]: In my draft text, I have:
`for (size_t j {}; j < 4; ++j)`
But of course, std::size() is better. Just double check your code matches your text.

File: Examples/NoModules/Chapter 08/Ex8_08.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  list_words(words);
+}
+
+void find_words(vector<string>& words, const string& text, const string& separators)
+{
+  size_t start {text.find_first_not_of(separators)};     // First word start index
+
+  while (start != string::npos)                          // Find the words
+  {
+    size_t end = text.find_first_of(separators, start + 1);  // Find end of word
```
Comment (MarcGregoire) [inline Examples/NoModules/Chapter 08/Ex8_08.cpp:32]: uniform init

File: Examples/NoModules/Chapter 08/Ex8_17.cpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  show_words(words);   // Output the words
+}
+
+void extract_words(Words& words, const std::string& text, const std::string& separators)
+{
+  size_t start {text.find_first_not_of(separators)}; // Start index of first word
+
+  while (start != std::string::npos)
+  {
+    size_t end = text.find_first_of(separators, start + 1); // Find end of a word
```
Comment (MarcGregoire) [inline Examples/NoModules/Chapter 08/Ex8_17.cpp:42]: uniform init

File: Exercises/NoModules/Chapter 08/Soln8_01.cpp [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+      months[date_month - 1], 
+      date_year
+    ) << std::endl;
+}
+
+// Reads an integer that is between lower and upper inclusive
+int validate_input(int lower, int upper, const std::string& description)
+{
+  int data {};
+  std::cout << "Please enter " << description << " from " << lower << " to " << upper << ": ";
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 08/Soln8_01.cpp:36]: std::format()?

File: Exercises/NoModules/Chapter 08/Soln8_05.cpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  }
+  
+  return true;
+}
+
+std::vector<unsigned> generateNumbers(unsigned to, unsigned from)
+{
+  std::vector<unsigned> result;
+  result.reserve(to - from + 1);
+  for (unsigned i = from; i <= to; ++i)
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 08/Soln8_05.cpp:76]: uniform init

File: Exercises/NoModules/Chapter 08/Soln8_05.cpp [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+*/
+
+int main()
+{
+  unsigned user_number{};
+  std::cout << "Would you be so kind as to enter a number? " << std::endl;
+  std::cin >> user_number;
+  
+  auto numbers = generateNumbers(user_number);
+  auto primes = filterPrimeNumbers(numbers);
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 08/Soln8_05.cpp:30]: twice uniform init

File: Exercises/NoModules/Chapter 08/Soln8_06.cpp [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+    sum += (number - average) * (number - average);
+  return std::sqrt(sum / numbers.size());
+}
+
+std::optional<double> computeVariance(const std::vector<unsigned>& numbers)
+{
+  if (numbers.empty())
+    return std::nullopt;
+
+  const double standardDeviation = *computeStandardDeviation(numbers);
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 08/Soln8_06.cpp:190]: uniform init

File: Exercises/NoModules/Chapter 08/Soln8_06.cpp [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+    return static_cast<double>(numbers[middleIndex] + numbers[middleIndex - 1]) / 2;
+  }
+}
+
+std::optional<double> computeStandardDeviation(const std::vector<unsigned>& numbers)
+{
+  if (numbers.empty())
+    return std::nullopt;
+
+  const double average = *computeAverage(numbers);
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 08/Soln8_06.cpp:178]: uniform init

File: Exercises/NoModules/Chapter 08/Soln8_06.cpp [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+void sort(std::vector<unsigned>& numbers, size_t start, size_t end)
+{
+  // start index must be less than end index for 2 or more elements
+  if (!(start < end))
+    return;
+
+  // Choose middle address to partition set
+  swap(numbers, start, (start + end) / 2);                            // Swap middle address with start
+
+                                    // Check words against chosen word
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 08/Soln8_06.cpp:98]: strange comment alignment

File: Exercises/NoModules/Chapter 08/Soln8_06.cpp [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+// Sort numbers in ascending sequence
+void sort(std::vector<unsigned>& numbers)
+{
+  if (!numbers.empty())
+    sort(numbers, 0, numbers.size() - 1);
+}
+
+void computeHighest(const std::vector<unsigned>& numbers, std::optional<unsigned>(&highest)[5])
+{
+  const auto numHighest = static_cast<int>(std::size(highest));
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 08/Soln8_06.cpp:121]: uniform init

File: Exercises/NoModules/Chapter 08/Soln8_06.cpp [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+    if (numberIndex >= 0)
+      highest[i] = numbers[numberIndex];
+    else
+      highest[i] = std::nullopt;
+  }
+}
+
+void computeLowest(const std::vector<unsigned>& numbers, std::optional<unsigned>(&lowest)[5])
+{
+  for (size_t i = 0; i < std::size(lowest); ++i)
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 08/Soln8_06.cpp:136]: uniform init

File: Exercises/NoModules/Chapter 08/Soln8_06.cpp [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+{
+  if (!numbers.empty())
+    sort(numbers, 0, numbers.size() - 1);
+}
+
+void computeHighest(const std::vector<unsigned>& numbers, std::optional<unsigned>(&highest)[5])
+{
+  const auto numHighest = static_cast<int>(std::size(highest));
+
+  for (int i = 0; i < numHighest; ++i)
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 08/Soln8_06.cpp:123]: uniform init

File: Exercises/NoModules/Chapter 08/Soln8_06.cpp [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+  {
+    const int numberIndex =
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 08/Soln8_06.cpp:125]: uniform init

File: Exercises/NoModules/Chapter 08/Soln8_06.cpp [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+  return average / numbers.size();
+}
+
+std::optional<double> computeMedian(const std::vector<unsigned>& numbers)
+{
+  if (numbers.empty())
+    return std::nullopt;
+
+  const auto numNumbers = numbers.size();
+  const auto middleIndex = numNumbers / 2;
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 08/Soln8_06.cpp:162]: twice uniform init

File: Exercises/NoModules/Chapter 08/Soln8_07A.cpp [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+     std::cout << "fib(" << i << ") = " << fib(i) << '\n';
+}
+
+unsigned long long fib(size_t n)
+{
+	// Initialise fib(i) and fib(i+1) for the first iteration of the loop where i == 0
+	unsigned long long fib_i{0};          // fib(i)   = fib(0) = 0
+	unsigned long long fib_i_1{1};        // fib(i+1) = fib(1) = 1
+	
+	for (size_t i = 0; i < n; ++i)
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 08/Soln8_07A.cpp:24]: uniform init

File: Exercises/NoModules/Chapter 08/Soln8_07A.cpp [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+	{
+		auto fib_i_2 = fib_i + fib_i_1;   // fib(i+2) = fib(i) + fib(i+1)
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 08/Soln8_07A.cpp:26]: uniform init

File: Exercises/NoModules/Chapter 08/Soln8_08.cpp [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+
+// Recursive function to calculate x to the power n
+long double power(double x, int n)
+{
+  if (n == 0)     return 1.0;
+  else if (n < 0) return 1.0 / power(x, -n);
+  else if (n % 2) return x * power(x, n-1);     // x is odd
+  
+  // If we make it this far, x > 0 and even
+  const auto y = power(x, n/2);
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 08/Soln8_08.cpp:24]: uniform init

File: Exercises/NoModules/Chapter 08/Soln8_08.cpp [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+
+  std::cout << std::endl;
+}
+
+// Recursive function to calculate x to the power n
+long double power(double x, int n)
+{
+  if (n == 0)     return 1.0;
+  else if (n < 0) return 1.0 / power(x, -n);
+  else if (n % 2) return x * power(x, n-1);     // x is odd
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 08/Soln8_08.cpp:21]: wrong comment, you're checking if n is odd

File: Exercises/NoModules/Chapter 08/Soln8_08.cpp [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+  
+  // If we make it this far, x > 0 and even
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 08/Soln8_08.cpp:23]: above you checked if n is even, not x

File: Exercises/NoModules/Chapter 08/Soln8_09.cpp [hunk 20]
```diff
@@ -0,0 +1,30 @@
+// Exercise 8_9 Count the number of multiplications performed by 
+// the divide and conquer power() function of Soln8_09.cpp
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 08/Soln8_09.cpp:2]: you mean Soln8_08.cpp

File: Exercises/NoModules/Chapter 08/Soln8_09.cpp [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+  std::cout << ++count << " multiplications" << std::endl;
+  return l * r;
+}
+
+// Recursive function to calculate x to the power n
+long double power(double x, int n)
+{
+  if (n == 0)     return 1.0;
+  else if (n < 0) return 1.0 / power(x, -n);
+  else if (n % 2) return mult(x, power(x, n-1));     // x is odd
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 08/Soln8_09.cpp:25]: same comment as before.

File: Exercises/NoModules/Chapter 08/Soln8_09.cpp [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+  
+  // If we make it this far, x > 0 and even
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 08/Soln8_09.cpp:27]: same comment as before.

File: Exercises/NoModules/Chapter 08/Soln8_09.cpp [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+  const auto y = power(x, n/2);
```
Comment (MarcGregoire) [inline Exercises/NoModules/Chapter 08/Soln8_09.cpp:28]: uniform init

**Review by MarcGregoire — CHANGES_REQUESTED** [review]

File: Examples/NoModules/Chapter 08/Ex8_06.cpp (hunk 1 above)
Comment (Peter-VanWeert) [reply Examples/NoModules/Chapter 08/Ex8_06.cpp:23]: It's std::size()

File: Exercises/NoModules/Chapter 08/Soln8_08.cpp (hunk 18 above)
Comment (Peter-VanWeert) [reply Exercises/NoModules/Chapter 08/Soln8_08.cpp:21]: Euhm, yes. I'm checking if n is odd, juts like the comments says...

File: Exercises/NoModules/Chapter 08/Soln8_08.cpp (hunk 19 above)
Comment (Peter-VanWeert) [reply Exercises/NoModules/Chapter 08/Soln8_08.cpp:23]: No. I checked that it is odd... If a number is even, n % 2 equals zero, and thus false, remember? ;-)

File: Exercises/NoModules/Chapter 08/Soln8_08.cpp (hunk 18 above)
Comment (MarcGregoire) [reply Exercises/NoModules/Chapter 08/Soln8_08.cpp]: but the comment says "x is odd" ...

File: Exercises/NoModules/Chapter 08/Soln8_08.cpp (hunk 19 above)
Comment (MarcGregoire) [reply Exercises/NoModules/Chapter 08/Soln8_08.cpp:23]: yes, but the comment says "x > 0 and even", shouldn't that be "n > 0 and even"?

**Pull request merged**